pkg/assembler/backends/testing/hashEqual.go

```diff
@@ -17,113 +17,201 @@ package testing
-			matchOrSkip = false
+	}
+	for _, m := range matchID {
+		if !slices.Contains(val, m) {
+			return false
+		}
+		val = slices.Delete(val, slices.Index(val, m), slices.Index(val, m)+1)
+	}
+	for _, m := range matchPartial {
+		match := false
+		remove := -1
+		for i, v := range val {
+			a, _ := c.artifactByID(v)
+			if (m.Algorithm == nil || strings.ToLower(*m.Algorithm) == a.algorithm) &&
+				(m.Digest == nil || strings.ToLower(*m.Digest) == a.digest) {
+				match = true
+				remove = i
+				break
+			}
 		}
-		if hashEqualSpec.Origin != nil && h.Origin != *hashEqualSpec.Origin {
-			matchOrSkip = false
+		if !match {
+			return false
 		}
+		val = slices.Delete(val, remove, remove+1)
+	}
+	return true
+}
 
-		if len(hashEqualSpec.Artifacts) > 0 && !filterEqualArtifact(h.Artifacts, hashEqualSpec.Artifacts) {
-			matchOrSkip = false
+// Query HashEqual
+
+func (c *demoClient) HashEqual(ctx context.Context, hSpec *model.HashEqualSpec) ([]*model.HashEqual, error) {
+	if len(hSpec.Artifacts) > 2 {
+		return nil, gqlerror.Errorf(
+			"HashEqual :: Provided spec has too many Artifacts")
+	}
+
+	// If ID is provided, try to look up, then check if rest matches
+	if hSpec.ID != nil {
+		id64, err := strconv.ParseUint(*hSpec.ID, 10, 32)
+		if err != nil {
+			return nil, gqlerror.Errorf("HashEqual :: invalid ID %s", err)
+		}
+		id := uint32(id64)
+		h, err := c.hashEqualByID(id)
+		if err != nil {
+			// Not found
+			return nil, nil
 		}
+		// If found by id, ignore rest of fields in spec and return as a match
+		return []*model.HashEqual{c.convHashEqual(h)}, nil
+	}
 
-		if matchOrSkip {
-			hashEquals = append(hashEquals, h)
+	var hashEquals []*model.HashEqual
+	// TODO if any artifacts are exact matches only search those backedges
+	for _, h := range c.hashEquals {
+		if noMatch(hSpec.Justification, h.justification) ||
+			noMatch(hSpec.Origin, h.origin) ||
+			noMatch(hSpec.Collector, h.collector) ||
+			!c.matchArtifacts(hSpec.Artifacts, h.artifacts) {
+			continue
 		}
+		hashEquals = append(hashEquals, c.convHashEqual(h))
 	}
 
 	return hashEquals, nil
 }
 
-func filterEqualArtifact(storedArtifacts []*model.Artifact, queryArtifacts []*model.ArtifactSpec) bool {
-	exists := make(map[model.Artifact]bool)
-	for _, value := range storedArtifacts {
-		exists[*value] = true
+func (c *demoClient) convHashEqual(h *hashEqualStruct) *model.HashEqual {
+	var artifacts []*model.Artifact
+	for _, id := range h.artifacts {
+		a, _ := c.artifactByID(id)
+		// TODO propagate error back
+		artifacts = append(artifacts, convArtifact(a))
 	}
-
-	// enforce lowercase for both the algorithm and digest when querying
-	for _, value := range queryArtifacts {
-		queryArt := model.Artifact{
-			Algorithm: strings.ToLower(*value.Algorithm),
-			Digest:    strings.ToLower(*value.Digest),
-		}
-		if _, ok := exists[queryArt]; ok {
-			return true
-		}
+	return &model.HashEqual{
+		ID:            fmt.Sprint(h.id),
```

Comment:
nit: can use `nodeID` for consistency where assigning IDs (for artifact, isOccurrence and isdependency)

Reply:
https://github.com/guacsec/guac/blob/main/pkg/assembler/backends/testing/backend.go#L133-L135